pkg/proc/mapiter.go

```diff
@@ -0,0 +1,603 @@
+	if !ok {
+		it.v.Unreadable = errSwissMapBadTableField
+		return
+	}
+	it.tableType, ok = tableptrtyp.Type.(*godwarf.StructType)
+	if !ok {
+		it.v.Unreadable = errSwissMapBadTableField
+		return
+	}
+	for _, field := range it.tableType.Field {
+		switch field.Name {
+		case "index":
+			it.tableFieldIndex = field
+		case "groups":
+			it.tableFieldGroups = field
+			groupstyp, ok := field.Type.(*godwarf.StructType)
+			if ok {
+				for _, field := range groupstyp.Field {
+					switch field.Name {
+					case "data":
+						it.groupsFieldData = field
+						typ, ok := field.Type.(*godwarf.PtrType)
+						if ok {
+							it.groupType, _ = resolveTypedef(typ.Type).(*godwarf.StructType)
+						}
+					case "lengthMask":
+						it.groupsFieldLengthMask = field
+					}
+				}
+			}
+		}
+	}
+	if it.groupType == nil || it.tableFieldIndex == nil || it.tableFieldGroups == nil || it.groupsFieldLengthMask == nil {
+		it.v.Unreadable = errSwissMapBadType
+		return
+	}
+	for _, field := range it.groupType.Field {
+		switch field.Name {
+		case "ctrl":
+			it.groupFieldCtrl = field
+		case "slots":
+			it.groupFieldSlots = field
+		}
+	}
+	if it.groupFieldCtrl == nil || it.groupFieldSlots == nil {
+		it.v.Unreadable = errSwissMapBadGroupTypeErr
+		return
+	}
+
+	slotsType, ok := resolveTypedef(it.groupFieldSlots.Type).(*godwarf.ArrayType)
+	if !ok {
+		it.v.Unreadable = errSwissMapBadGroupTypeErr
+		return
+	}
+	slotType, ok := slotsType.Type.(*godwarf.StructType)
+	if !ok {
+		it.v.Unreadable = errSwissMapBadGroupTypeErr
+		return
+	}
+	for _, field := range slotType.Field {
+		switch field.Name {
+		case "key":
+			it.slotFieldKey = field
+		case "elem":
+			it.slotFieldElem = field
+		}
+	}
+	if it.slotFieldKey == nil || it.slotFieldElem == nil {
+		it.v.Unreadable = errSwissMapBadGroupTypeErr
+		return
+	}
+
+	if it.dirLen <= 0 {
+		// small maps, convert it.dirPtr to be of type *group, then dereference it
+		it.dirPtr.DwarfType = pointerTo(fakeArrayType(1, it.groupType), it.v.bi.Arch)
+		it.dirPtr.RealType = it.dirPtr.DwarfType
+		it.dirPtr = it.dirPtr.maybeDereference()
+		it.dirLen = 1
+		it.tab = &swissTable{groups: it.dirPtr} // so that we don't try to load this later on
+		return
+	}
+
+	// normal map, convert it.dirPtr to be of type *[dirLen]*table, then dereference it
+
+	it.dirPtr.DwarfType = pointerTo(fakeArrayType(uint64(it.dirLen), tableptrtyp), it.v.bi.Arch)
+	it.dirPtr.RealType = it.dirPtr.DwarfType
+	it.dirPtr = it.dirPtr.maybeDereference()
+}
+
+// derived from $GOROOT/src/internal/runtime/maps/table.go and $GOROOT/src/runtime/runtime-gdb.py
+func (it *mapIteratorSwiss) next() bool {
+	if it.v.Unreadable != nil {
+		return false
+	}
+	for it.dirIdx < it.dirLen {
+		if it.tab == nil {
+			it.loadCurrentTable()
+			if it.tab == nil {
+				return false
+			}
+			if it.tab.index != it.dirIdx {
+				it.nextTable()
+				continue
+			}
+		}
+
+		for ; it.groupIdx <= uint64(it.tab.groups.Len); it.nextGroup() {
```

Comment:
it.tab.groups.Len is the array length, right? Thus shouldn't this be < not <=?

Reply:
You're right.